From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=2E=20Neusch=C3=A4fer?= <j.neuschaefer@gmx.net>
Date: Thu, 21 Dec 2023 23:45:45 +0100
Subject: [PATCH] 6040: ARM: Fix alignment requirement for switch8_r3 patterns

While analyzing a program in which the switch8_r3 pattern occurs,
the pattern was not detected, and thus switch statements could not
be decompiled properly.

The cause turned out to be that the alignment requirement
`<align mark="0" bits="3"/>` means that 8-byte aligment is required,
but the switch8_r3 function in this program was only 4-byte aligned.

This commit adjusts the alignment requirements listed for switch8_r3
in ARM_switch_patterns.xml to permit 4-byte alignment.
---
 Ghidra/Processors/ARM/data/patterns/ARM_switch_patterns.xml | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/patterns/ARM_switch_patterns.xml b/Ghidra/Processors/ARM/data/patterns/ARM_switch_patterns.xml
index f622f10fce..65b4de51f3 100644
--- a/Ghidra/Processors/ARM/data/patterns/ARM_switch_patterns.xml
+++ b/Ghidra/Processors/ARM/data/patterns/ARM_switch_patterns.xml
@@ -94,7 +94,7 @@
            add        ip,lr,r3, lsl #0x1   |   add lr,lr,r3, lsl #0x1
            bx         ip                   |   bx lr
        -->
-      <align mark="0" bits="3"/>
+      <align mark="0" bits="2"/>
       <setcontext name="TMode" value="0"/>
       <funcstart label="switch8_r3"/>
   </pattern>
@@ -108,7 +108,7 @@
            add        ip,lr,r3, lsl #0x1   |   add lr,lr,r3, lsl #0x1
            bx         ip                   |   bx lr
        -->
-      <align mark="0" bits="3"/>
+      <align mark="0" bits="2"/>
       <setcontext name="TMode" value="0"/>
       <funcstart label="switch8_r3"/>
   </pattern>
-- 
2.45.1

